File: otto-mobile/otto-android-LEGACY-DO-NOT-OPEN/app/src/main/java/to/ottomot/driftd/routebuilder/engine/RoutePolylineTurnExtractor.kt
package to.ottomot.driftd.routebuilder.engine

import kotlin.math.PI
import kotlin.math.abs
import to.ottomot.driftd.core.event.haversineMeters

/**
 * Derives turn-like coordinates from a saved road polyline without re-calling directions.
 */
object RoutePolylineTurnExtractor {
    private const val MINIMUM_TURN_ANGLE_DEGREES = 28.0
    private const val MINIMUM_ARC_SPACING_METERS = 40.0
    private const val FINISH_BUFFER_METERS = RouteAutoPathPointGenerator.Options.FINISH_BUFFER_METERS

    fun turnCoordinates(along: List<RouteLatLng>): List<RouteLatLng> {
        if (along.size < 3) return emptyList()

        val totalLength = RoutePolylineGeometry.polylineTotalLength(along)
        if (totalLength <= FINISH_BUFFER_METERS * 2) return emptyList()

        val sampleStride = maxOf(1, along.size / 500)
        val cumulative = ArrayList<Double>(along.size)
        cumulative.add(0.0)
        for (segmentIndex in 0 until along.size - 1) {
            val start = along[segmentIndex]
            val end = along[segmentIndex + 1]
            cumulative.add(
                cumulative[segmentIndex] +
                    haversineMeters(start.lat, start.lng, end.lat, end.lng),
            )
        }

        val results = mutableListOf<RouteLatLng>()
        var lastTurnArcLength = -Double.MAX_VALUE
        var index = sampleStride

        while (index < along.size - sampleStride) {
            val previous = along[index - sampleStride]
            val current = along[index]
            val next = along[index + sampleStride]

            val incoming = RoutePolylineGeometry.bearingBetween(from = previous, to = current)
            val outgoing = RoutePolylineGeometry.bearingBetween(from = current, to = next)
            val turnAngle = normalizedAngleDelta(startDegrees = incoming, endDegrees = outgoing)

            if (turnAngle >= MINIMUM_TURN_ANGLE_DEGREES) {
                val arcLength = cumulative[index]
                if (arcLength > FINISH_BUFFER_METERS &&
                    arcLength < totalLength - FINISH_BUFFER_METERS &&
                    arcLength - lastTurnArcLength >= MINIMUM_ARC_SPACING_METERS
                ) {
                    results.add(current)
                    lastTurnArcLength = arcLength
                }
            }

            index += sampleStride
        }

        return results
    }

    private fun normalizedAngleDelta(startDegrees: Double, endDegrees: Double): Double {
        var delta = abs(endDegrees - startDegrees) % 360.0
        if (delta > 180.0) delta = 360.0 - delta
        return delta
    }
}

File: otto-mobile/otto-mobile/RoutePolylineTurnExtractor.swift
import CoreLocation
import Foundation
import MapKit

/// Derives turn-like coordinates from a saved road polyline without re-calling directions on editor open.
enum RoutePolylineTurnExtractor {
    private static let minimumTurnAngleDegrees = 28.0
    private static let minimumArcSpacingMeters = 40.0
    private static let finishBufferMeters = RouteAutoPathPointGenerator.Options.finishBufferMeters

    static func turnCoordinates(along roadCoordinates: [CLLocationCoordinate2D]) -> [CLLocationCoordinate2D] {
        guard roadCoordinates.count >= 3 else { return [] }

        let totalLength = RouteMapGeometry.polylineTotalLength(roadCoordinates)
        guard totalLength > finishBufferMeters * 2 else { return [] }

        let sampleStride = max(1, roadCoordinates.count / 500)
        var cumulative: [Double] = [0]
        cumulative.reserveCapacity(roadCoordinates.count)
        for segmentIndex in 0..<(roadCoordinates.count - 1) {
            let start = MKMapPoint(roadCoordinates[segmentIndex])
            let end = MKMapPoint(roadCoordinates[segmentIndex + 1])
            cumulative.append(cumulative[segmentIndex] + start.distance(to: end))
        }

        var results: [CLLocationCoordinate2D] = []
        var lastTurnArcLength = -Double.greatestFiniteMagnitude
        var index = sampleStride

        while index < roadCoordinates.count - sampleStride {
            let previous = roadCoordinates[index - sampleStride]
            let current = roadCoordinates[index]
            let next = roadCoordinates[index + sampleStride]

            let incoming = RouteMapGeometry.bearingBetween(from: previous, to: current)
            let outgoing = RouteMapGeometry.bearingBetween(from: current, to: next)
            let turnAngle = normalizedAngleDelta(from: incoming, to: outgoing)

            if turnAngle >= minimumTurnAngleDegrees {
                let arcLength = cumulative[index]
                if arcLength > finishBufferMeters,
                   arcLength < totalLength - finishBufferMeters,
                   arcLength - lastTurnArcLength >= minimumArcSpacingMeters {
                    results.append(current)
                    lastTurnArcLength = arcLength
                }
            }

            index += sampleStride
        }

        return results
    }

    private static func normalizedAngleDelta(from startDegrees: Double, to endDegrees: Double) -> Double {
        var delta = abs(endDegrees - startDegrees).truncatingRemainder(dividingBy: 360)
        if delta > 180 { delta = 360 - delta }
        return delta
    }
}
